File: skills/azure-resource-abbreviations/scripts/generate_template.py
#!/usr/bin/env python3
"""
Generate a naming template for a specific Azure resource type.
Usage: python generate_template.py "Virtual Machine"
"""

import sys
from pathlib import Path
import re

# Add parent directory to path
sys.path.insert(0, str(Path(__file__).parent))

from azure_abbreviations import get_abbreviation, search_abbreviation


NAMING_TEMPLATES = {
    "standard": "{abbr}-{env}-{instance}-{region}",
    "simple": "{abbr}{env}{instance}",
    "descriptive": "{abbr}-{purpose}-{env}-{instance}",
}

ENVIRONMENT_CODES = {
    "dev": "Development",
    "test": "Testing",
    "stg": "Staging",
    "prod": "Production",
}

REGION_CODES = {
    "eus": "East US",
    "wus": "West US",
    "neu": "North Europe",
    "weu": "West Europe",
    "sea": "Southeast Asia",
    "eas": "East Asia",
    "jpeast": "Japan East",
}


def generate_examples(abbreviation: str) -> dict:
    """Generate naming examples for an abbreviation."""
    examples = {}
    
    # Standard template
    examples["standard"] = {
        "template": NAMING_TEMPLATES["standard"],
        "example": f"{abbreviation}-prod-web01-eus",
        "description": "Resource-Environment-Instance-Region"
    }
    
    # Simple template
    examples["simple"] = {
        "template": NAMING_TEMPLATES["simple"],
        "example": f"{abbreviation}prod1",
        "description": "ResourceEnvironmentInstance (no separators)"
    }
    
    # Descriptive template
    examples["descriptive"] = {
        "template": NAMING_TEMPLATES["descriptive"],
        "example": f"{abbreviation}-webserver-prod-01",
        "description": "Resource-Purpose-Environment-Instance"
    }
    
    return examples


def main():
    if len(sys.argv) < 2:
        print("Usage: python generate_template.py <resource_type>")
        print("\nExamples:")
        print("  python generate_template.py 'Virtual Machine'")
        print("  python generate_template.py 'Web App'")
        sys.exit(1)
    
    resource_type = " ".join(sys.argv[1:])
    
    # Try exact match first
    abbreviation = get_abbreviation(resource_type)
    if not abbreviation:
        # Try search
        results = search_abbreviation(resource_type)
        if results:
            first_category = list(results.keys())[0]
            first_resource = list(results[first_category].keys())[0]
            abbreviation = results[first_category][first_resource]
            print(f"Using '{first_resource}': {abbreviation}\n")
        else:
            print(f"✗ Resource type '{resource_type}' not found")
            sys.exit(1)
    
    print(f"Azure Resource: {resource_type}")
    print(f"Abbreviation: {abbreviation}\n")
    print("=" * 70)
    print("NAMING TEMPLATES\n")
    
    examples = generate_examples(abbreviation)
    
    for template_name, details in examples.items():
        print(f"[{template_name.upper()}]")
        print(f"Template:     {details['template']}")
        print(f"Example:      {details['example']}")
        print(f"Description:  {details['description']}")
        print()
    
    print("=" * 70)
    print("\nNAMING COMPONENTS\n")
    print("Environment Codes:")
    for code, meaning in ENVIRONMENT_CODES.items():
        print(f"  {code:<6} = {meaning}")
    
    print("\nCommon Region Codes:")
    for code, meaning in REGION_CODES.items():
        print(f"  {code:<8} = {meaning}")
    
    print("\n" + "=" * 70)
    print("\nBEST PRACTICES:")
    print("  • Use lowercase letters and numbers")
    print("  • Use hyphens (-) as separators")
    print("  • Keep names concise but descriptive")
    print("  • Include environment and region for production resources")
    print("  • Use consistent abbreviations across your organization")
    print("  • Respect character limits (varies by resource type)")


if __name__ == "__main__":
    main()

File: azure-resource-abbreviations/scripts/generate_template.cs
#:property PublishAot=false
#:property NoWarn=CA2266
#:include AzureResourceAbbreviationsSupport.cs

var namingTemplates = new Dictionary<string, (string Template, string Example, string Description)>(StringComparer.Ordinal)
{
    ["standard"] = ("{abbr}-{env}-{instance}-{region}", "{abbr}-prod-web01-eus", "Resource-Environment-Instance-Region"),
    ["simple"] = ("{abbr}{env}{instance}", "{abbr}prod1", "ResourceEnvironmentInstance (no separators)"),
    ["descriptive"] = ("{abbr}-{purpose}-{env}-{instance}", "{abbr}-webserver-prod-01", "Resource-Purpose-Environment-Instance")
};

var environmentCodes = new Dictionary<string, string>(StringComparer.Ordinal)
{
    ["dev"] = "Development",
    ["test"] = "Testing",
    ["stg"] = "Staging",
    ["prod"] = "Production"
};

var regionCodes = new Dictionary<string, string>(StringComparer.Ordinal)
{
    ["eus"] = "East US",
    ["wus"] = "West US",
    ["neu"] = "North Europe",
    ["weu"] = "West Europe",
    ["sea"] = "Southeast Asia",
    ["eas"] = "East Asia",
    ["jpeast"] = "Japan East"
};

if (args.Length == 0)
{
    Console.WriteLine("Usage: dotnet run --file scripts\\generate_template.cs -- <resource_type>");
    Console.WriteLine();
    Console.WriteLine("Examples:");
    Console.WriteLine("  dotnet run --file scripts\\generate_template.cs -- \"Virtual Machine\"");
    Console.WriteLine("  dotnet run --file scripts\\generate_template.cs -- webSitesAppService");
    return 1;
}

var query = string.Join(' ', args);
var entry = AzureResourceAbbreviationsSupport.FindExact(query);
if (entry is null)
{
    entry = AzureResourceAbbreviationsSupport.Search(query).FirstOrDefault();
}

if (entry is null)
{
    Console.WriteLine($"Resource type '{query}' not found.");
    Console.WriteLine($"Reference file: {AzureResourceAbbreviationsSupport.GetReferencePath()}");
    return 1;
}

Console.WriteLine($"Display name:    {entry.DisplayName}");
Console.WriteLine($"Resource key:    {entry.ResourceTypeKey}");
Console.WriteLine($"Category:        {entry.Category}");
Console.WriteLine($"Official prefix: {entry.OfficialPrefix}");
Console.WriteLine($"Naming token:    {entry.NamingToken}");
Console.WriteLine();
Console.WriteLine(new string('=', 70));
Console.WriteLine("NAMING TEMPLATES");
Console.WriteLine();

foreach (var (templateName, details) in namingTemplates)
{
    Console.WriteLine($"[{templateName.ToUpperInvariant()}]");
    Console.WriteLine($"Template:     {details.Template}");
    Console.WriteLine($"Example:      {details.Example.Replace("{abbr}", entry.NamingToken, StringComparison.Ordinal)}");
    Console.WriteLine($"Description:  {details.Description}");
    Console.WriteLine();
}

Console.WriteLine(new string('=', 70));
Console.WriteLine();
Console.WriteLine("NAMING COMPONENTS");
Console.WriteLine();
Console.WriteLine("Environment Codes:");
foreach (var (code, meaning) in environmentCodes)
{
    Console.WriteLine($"  {code,-6} = {meaning}");
}

Console.WriteLine();
Console.WriteLine("Common Region Codes:");
foreach (var (code, meaning) in regionCodes)
{
    Console.WriteLine($"  {code,-8} = {meaning}");
}

Console.WriteLine();
Console.WriteLine(new string('=', 70));
Console.WriteLine();
Console.WriteLine("BEST PRACTICES:");
Console.WriteLine("  - Use lowercase letters and numbers.");
Console.WriteLine("  - Use hyphens (-) as separators unless the resource type forbids them.");
Console.WriteLine("  - Keep names concise but descriptive.");
Console.WriteLine("  - Include environment and region for production resources.");
Console.WriteLine("  - Use the JSON reference as the authoritative source for prefixes.");
return 0;
